src/java/org/apache/cassandra/config/DataStorageSpec.java

```diff
@@ -38,117 +36,111 @@
                                              " where case matters and only non-negative values are accepted");
         }
     }
-    /**
-     * Creates a {@code DataStorageSpec} of the specified amount of bytes.
-     *
-     * @param bytes the amount of bytes
-     * @return a {@code DataStorageSpec}
-     */
-    public static DataStorageSpec inBytes(long bytes)
+
+    private DataStorageSpec(String value, DataStorageUnit minUnit, long max)
     {
-        return new DataStorageSpec(bytes, DataStorageUnit.BYTES);
+        this(value, minUnit);
+
+        validateMinUnit(unit, minUnit, value);
+        validateQuantity(value, quantity(), unit(), minUnit, max);
     }
 
-    /**
-     * Creates a {@code DataStorageSpec} of the specified amount of kibibytes.
-     *
-     * @param kibibytes the amount of kibibytes
-     * @return a {@code DataStorageSpec}
-     */
-    public static DataStorageSpec inKibibytes(long kibibytes)
+    private void validateMinUnit(DataStorageUnit unit, DataStorageUnit minUnit, String value)
     {
-        return new DataStorageSpec(kibibytes, KIBIBYTES);
+        if (unit.compareTo(minUnit) < 0)
+            throw new ConfigurationException(String.format("Invalid data storage: %s Accepted units:%s", value, acceptedUnits(minUnit)));
     }
 
-    /**
-     * Creates a {@code DataStorageSpec} of the specified amount of mebibytes.
-     *
-     * @param mebibytes the amount of mebibytes
-     * @return a {@code DataStorageSpec}
-     */
-    public static DataStorageSpec inMebibytes(long mebibytes)
+    private String acceptedUnits(DataStorageUnit minUnit)
+    {
+        DataStorageUnit[] units = DataStorageUnit.values();
+        return Arrays.toString(Arrays.copyOfRange(units, minUnit.ordinal(), units.length));
+    }
+
+    private static void validateQuantity(String value, long quantity, DataStorageUnit sourceUnit, DataStorageUnit minUnit, long max)
     {
-        return new DataStorageSpec(mebibytes, MEBIBYTES);
+        // no need to validate for negatives as they are not allowed at first place from the regex
+
+        if (minUnit.convert(quantity, sourceUnit) >= max)
+            throw new ConfigurationException("Invalid data storage: " + value + ". It shouldn't be more than " +
+                                             (max - 1) + " in " + minUnit.name().toLowerCase());
     }
 
+    private static void validateQuantity(long quantity, DataStorageUnit unit, DataStorageUnit minUnit, long max)
+    {
+        if (quantity < 0)
+            throw new ConfigurationException("Invalid data storage: value must be non-negative");
+
+        if (minUnit.convert(quantity, unit) >= max)
+            throw new ConfigurationException("Invalid data storage: " + quantity + " " + unit.name().toLowerCase() + ". It shouldn't be more than " +
+                                             (max - 1) + " in " + minUnit.name().toLowerCase());
```

Comment:
Perhaps this message is long enough to use String.format to increase readability:
```suggestion
            throw new ConfigurationException(String.format("Invalid data storage: %d %s. It shouldn't be more than %d in %s",
                                                           quantity, unit.name().toLowerCase(),
                                                           max - 1, minUnit.name().toLowerCase()));
```